src/loan/terms/simple/loan/pwn_simple_loan.cairo

```diff
@@ -81,100 +146,704 @@ mod PwnSimpleLoan {
+                self.domain_separator.read(), starknet::get_contract_address().try_into().unwrap()
+            ];
+            let domain_seperator_hash = poseidon_hash_span(hash_elements.span());
+
+            let hash_elements: Array<felt252> = array![
+                1901,
+                domain_seperator_hash,
+                EXTENSION_PROPOSAL_TYPEHASH,
+                extension.loan_id,
+                extension.compensation_address.try_into().unwrap(),
+                extension.compensation_amount.try_into().unwrap(),
+                extension.duration.into(),
+                extension.expiration.into(),
+                extension.proposer.try_into().unwrap(),
+                extension.nonce_space,
+                extension.nonce
+            ];
+            poseidon_hash_span(hash_elements.span())
         }
 
         fn get_loan(self: @ContractState, loan_id: felt252) -> GetLoanReturnValue {
-            Default::default()
+            let loan = self.loans.read(loan_id);
+            let loan_owner: ContractAddress = if (loan.status != 0) {
+                ERC721ABIDispatcher { contract_address: self.loan_token.read().contract_address }
+                    .owner_of(loan_id.try_into().unwrap())
+            } else {
+                Default::default()
+            };
+            let loan_return_value = GetLoanReturnValue {
+                status: self._get_loan_status(loan_id),
+                start_timestamp: loan.start_timestamp,
+                default_timestamp: loan.default_timestamp,
+                borrower: loan.borrower,
+                original_lender: loan.original_lender,
+                loan_owner,
+                accruing_interest_APR: loan.accruing_interest_APR,
+                fixed_interest_amount: loan.fixed_interest_amount,
+                credit: ERC20(loan.credit_address, loan.principal_amount),
+                collateral: loan.collateral,
+                original_source_of_funds: loan.original_source_of_funds,
+                repayment_amount: self.get_loan_repayment_amount(loan_id)
+            };
+            loan_return_value
         }
 
         fn get_is_valid_asset(self: @ContractState, asset: Asset) -> bool {
-            true
+            asset.is_valid(Option::Some(self.category_registry.read().contract_address))
         }
 
         fn get_loan_metadata_uri(self: @ContractState) -> ByteArray {
-            Default::default()
+            let this_contract = starknet::get_contract_address();
+            self.config.read().loan_metadata_uri(this_contract)
         }
 
-        fn get_state_fingerprint(self: @ContractState) -> felt252 {
-            0
+        fn get_state_fingerprint(self: @ContractState, token_id: felt252) -> felt252 {
+            let loan = self.loans.read(token_id);
+            if (loan.status == 0) {
+                return 0;
+            }
+
+            // @note: here fixed_interest_amount is in u256, shall we break it to low and high?
```

Comment:
Yes good idea to break it down instead of try_into() so we avoid overflow in the rare cases that the value is > than felt252